lib/msal-browser/src/cache/BrowserStorage.ts

```diff
@@ -0,0 +1,239 @@
+        const msalKey = this.generateCacheKey(key, true);
+        this.windowStorage.removeItem(msalKey);
+        if (this.rollbackEnabled) {
+            this.windowStorage.removeItem(this.generateCacheKey(key, false));
+        }
+        if (this.cacheConfig.storeAuthStateInCookie) {
+            this.clearItemCookie(msalKey);
+        }
+    }
+    
+    containsKey(key: string): boolean {
+        const msalKey = this.generateCacheKey(key, true);
+        return this.windowStorage.hasOwnProperty(msalKey) || this.windowStorage.hasOwnProperty(key);
+    }
+    
+    getKeys(): string[] {
+        return Object.keys(this.windowStorage);
+    }
+
+    clear(): void {
+        let key: string;
+        for (key in this.windowStorage) {
+            // Check if key contains msal prefix; For now, we are clearing all the cache items created by MSAL.js
+            if (this.windowStorage.hasOwnProperty(key) && (key.indexOf(Constants.CACHE_PREFIX) !== -1) && (key.indexOf(this.clientId) !== -1)) {
+                this.removeItem(key);
+            }
+        }
+    }
+
+    /**
+     * Reset all temporary cache items
+     * @param state 
+     */
+    resetTempCacheItems(state: string): void {
+        const storage = window[this.cacheConfig.cacheLocation];
+        let key: string;
+        // check state and remove associated cache
+        for (key in storage) {
+            if (!state || key.indexOf(state) !== -1) {
+                const splitKey = key.split(Constants.RESOURCE_DELIM);
+                const keyState = splitKey.length > 1 ? splitKey[splitKey.length-1]: null;
+                if (keyState === state && !this.tokenRenewalInProgress(keyState)) {
+                    this.removeItem(key);
+                    this.setItemCookie(key, "", -1);
+                    this.clearMsalCookie(state);
+                }
+            }
+        }
+        // delete the interaction status cache
+        this.removeItem(TemporaryCacheKeys.INTERACTION_STATUS);
+        this.removeItem(TemporaryCacheKeys.REDIRECT_REQUEST);
+    }
+
+    /**
+     * Return if the token renewal is still in progress
+     * @param stateValue
+     */
+    private tokenRenewalInProgress(stateValue: string): boolean {
+        const renewStatus = this.getItem(`${TemporaryCacheKeys.RENEW_STATUS}|${stateValue}`);
+        return !!(renewStatus && renewStatus === BrowserConstants.inProgress);
+    }
+
+    /**
+     * add value to cookies
+     * @param cName
+     * @param cValue
+     * @param expires
+     */
+    setItemCookie(cName: string, cValue: string, expires?: number): void {
+        let cookieStr = cName + "=" + cValue + ";";
+        if (expires) {
+            const expireTime = this.getCookieExpirationTime(expires);
+            cookieStr += "expires=" + expireTime + ";";
+        }
+
+        document.cookie = cookieStr;
+    }
+
+    /**
+     * get one item by key from cookies
+     * @param cName
+     */
+    getItemCookie(cName: string): string {
```

Comment:
I think this function could use more descriptive variables. I'm assuming `cName` is canonical name, and what are `c` and `ca` in this context?

Reply:
Or is it cookie name? maybe this is just my unfamiliarity, lol

Reply:
I will go through and add more description.

Reply:
Done.